### PR TITLE
HDDS-6634. Incorrect datanode state drop after lose connection to SCM server

### PR DESCRIPTION
**Pull request by sadanand48**
## What changes were proposed in this pull request?
Change endpoint state to GetVersion in case of Re-Register to prevent cases where older version with older cluster ID of datanode registers back.

To add more context : 
Datanode has an EndpointStateMachine which undergoes various states in sequence - {GETVERSION,REGISTER,HEARTBEAT,SHUTDOWN}
When DN is started it first gets version from SCM and creates its storage dirs (adds version files etc) and then registers with SCM. Now that it is registered, it will keep sending HeartBeats and EndpointState = HEARTBEAT. If during this time, SCM is re-initialised i.e all prev storage dirs are deleted and SCM is inited with new cluster ID then DN still keeps sending Heartbeats and re-registers . In this case if version is not checked an old DN can register to  a new SCM and cause issues which is why the PR adds a check to validate clusterID's in case of Re-Register.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-6634

## How was this patch tested?
Unit test

### CODE REVIEW

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/states/endpoint/HeartbeatEndpointTask.java [hunk 1]
```diff
@@ -452,7 +452,7 @@ private void processReregisterCommand() {
         LOG.debug("Received SCM notification to register."
             + " Interrupt HEARTBEAT and transit to REGISTER state.");
```
Comment (aryangupta1998) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/states/endpoint/HeartbeatEndpointTask.java:453]: Can we update the log to the GETVERSION state?
```suggestion
            + " Interrupt HEARTBEAT and transit to GETVERSION state.");
```

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/scm/TestStorageContainerManager.java [hunk 2]
```diff
@@ -370,6 +375,78 @@ public void testBlockDeletionTransactions() throws Exception {
...
+          datanode.getDatanodeDetails());
+      GenericTestUtils.waitFor(
+          () -> scmDnHBDispatcherLog.getOutput().contains(expectedLog), 100,
+          3000);
+      ExitUtil.disableSystemExit();
+      // As part of processing response for re-register, DN EndpointStateMachine
+      // goes to GET-VERSION state which checks if there is already existing
+      // version file on the DN & if the clusterID matches with that of the SCM
+      // In this case, it won't match and gets InconsistentStorageStateException
+      // and DN shuts down.
```
Comment (aryangupta1998) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/scm/TestStorageContainerManager.java:437]: After this comment can we also check the HeartbeatEndpointTask log that DN has received a response from SCM to re-register itself and as part of it DN EndpointStateMachine goes to GETVERSION state and then DN shuts down incase of a clusterID mismatch, maybe something like,
```suggestion
      GenericTestUtils.LogCapturer heartbeatEndpointTaskLog =
              GenericTestUtils.LogCapturer.captureLogs(HeartbeatEndpointTask.LOG);
       String expectedLog2 = "Received SCM notification to register."
              + " Interrupt HEARTBEAT and transit to GETVERSION state.";
      GenericTestUtils.waitFor(
              () -> heartbeatEndpointTaskLog.getOutput().contains(expectedLog2), 100,
              3000);       
```

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/states/endpoint/HeartbeatEndpointTask.java (hunk 1 above)
Comment (sadanand48) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/states/endpoint/HeartbeatEndpointTask.java]: Done.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/scm/TestStorageContainerManager.java (hunk 2 above)
Comment (sadanand48) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/scm/TestStorageContainerManager.java:443]: Done.

**Comment by adoroszlai** [comment]
@Pochatkin would you like to review?

**Review by sumitagrawl — APPROVED** [review]
@sadanand48 Thanks for working over this, LGTM +1

**Comment by adoroszlai** [comment]
Thanks @sadanand48 for the patch, @aryangupta1998, @sumitagrawl for the review.

**Pull request merged**